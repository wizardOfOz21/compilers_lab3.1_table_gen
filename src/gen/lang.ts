import {
    ExternalVertex,
    InternalVertex,
    isExternal,
    isInternal,
    ParseTable,
    Rule,
    Vertex,
} from "../parse";
import { getTable } from "../utils";
import { getFirst } from "./first";
import { getFollow } from "./follow";

export type Language = {
    terms: string[];
    nterms: string[];
    rules: Rule[];
    axioms: string[];
};

let terms: string[];
let nterms: string[];
let rules: Rule[];
let axioms: string[];

export const toLang = (root: InternalVertex): Language => {

    terms = [];
    nterms = [];
    rules = [];
    axioms = [];

    dfs(root, toLangV);

    return { terms, nterms, rules, axioms };
};

const dfs = (v: Vertex, visit: (v: Vertex) => void) => {
    visit(v);
    if (isExternal(v)) {
        return;
    }
    for (const c of v.childs) {
        dfs(c, visit);
    }
};

const getNterms = (v: Vertex) => {
    if (isExternal(v)) {
        return;
    }
    switch (v.nterm) {
        case "IL":
            nterms.push((v.childs[0] as ExternalVertex).val.attr);
            break;

        case "ILt":
            if (v.childs.length !== 0)
                nterms.push((v.childs[1] as ExternalVertex).val.attr);
            break;
    }
};

const getTerms = (v: Vertex) => {
    if (isExternal(v)) {
        return;
    }
    if (v.nterm === "E") {
        terms.push((v.childs[0] as ExternalVertex).val.attr);
    }
};

const getRules = (v: InternalVertex) => {
    const lhs = (v.childs[0] as ExternalVertex).val.attr;

    dfs(v, (s: Vertex) => {
        if (isExternal(s)) {
            return;
        }
        if (s.nterm === "S") {
            const rhs = [];
            dfs(s, (e: Vertex) => {
                if (isExternal(e)) {
                    return;
                }
                if (e.nterm === "E") {
                    rhs.push((e.childs[0] as ExternalVertex).val.attr);
                }
            });
            rules.push({ lhs, rhs });
        }
    });
};

const toLangV = (v: Vertex) => {
    if (isExternal(v)) {
        return;
    }
    switch (v.nterm) {
        case "NTS":
            dfs(v, getNterms);
            return;
        case "TS":
            dfs(v, getTerms);
            return;
        case "R":
            getRules(v);
            return;
        case "AX":
            axioms.push((v.childs[1] as ExternalVertex).val.attr);
        default:
            break;
    }
};

export const getParseTable = (lang: Language): ParseTable => {
    const [FIRST, f] = getFirst(lang);
    const [FOLLOW, fl] = getFollow(lang, FIRST);

    lang.terms.push('$');

    const table: ParseTable = getTable(lang.nterms.length, lang.terms.length, 'error');
    const add = (X,a,r) => {
        const pred = table[lang.nterms.indexOf(X)][lang.terms.indexOf(a)];
        if (!pred) {return;}
        if (pred !== 'error') {throw new Error("Not LL1 grammar ):"); }

        table[lang.nterms.indexOf(X)][lang.terms.indexOf(a)] = r;
    }

    lang.rules.forEach(r => {
        const X = r.lhs;
        const u = r.rhs;

        for (const a of FIRST(u)) {
            add(X,a,r);
        }

        if (FIRST(u).has('epsilon')) {
            for (const b of FOLLOW(X)) {
                add(X,b,r);
            }
        }
    })

    return table;
}
