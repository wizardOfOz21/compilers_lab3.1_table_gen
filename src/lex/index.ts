// re должно начинаться с ^, не должно содержать флага g
export type DomainSpec = {
    tag: string;
    re: RegExp;
    getAttr?: (str: RegExpMatchArray) => any;
};

export type Token = {
    tag: string;
    attr: any;
};

const EOF: DomainSpec = {
    tag: "EOF",
    re: /^\$/,
};

const NL: DomainSpec = {
    tag: "NL",
    re: /^\n/,
};

const SP: DomainSpec = {
    tag: "SP",
    re: /^[ ]+/,
};

// input – строка терминированная $
export function lex(input: string, specs: DomainSpec[]): Token[] {
    let col = 1;
    let row = 1;
    const tokens: Token[] = [];
    specs.push(EOF);
    specs.push(NL);
    specs.push(SP);
    while (true) {
        let hasMatch = false;
        for (const spec of specs) {
            const match = input.match(spec.re);
            if (match) {
                const matchLength = match[0].length;
                hasMatch = true;
                switch (spec.tag) {
                    case EOF.tag:
                        tokens.push({
                            tag: '$',
                            attr: null,
                        });
                        return tokens;
                    case NL.tag:
                        row += 1;
                        col = 1;
                        input = input.slice(matchLength);
                        break;
                    case SP.tag:
                        col += matchLength;
                        input = input.slice(matchLength);
                        break;
                    default:
                        col += matchLength;
                        input = input.slice(matchLength);
                        tokens.push({
                            tag: spec.tag,
                            attr: spec.getAttr ? spec.getAttr(match) : null,
                        });
                        break;
                }
                break;
            }
        }
        if (!hasMatch) {
            throw new Error(`unknown lexem on ${row}.${col}`);
        }
    }
}
