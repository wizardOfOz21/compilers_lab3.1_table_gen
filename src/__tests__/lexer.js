import { lex } from "../lex";
import { domains } from "../lex/specs/lang";

test("lexer работаетб", () => {
    const input =
        "abba ::= ; , || 'asdasd' axiom terminal epsilon non-terminal # asdasd 1$";
    expect(lex(input, domains)).toMatchSnapshot();
});
