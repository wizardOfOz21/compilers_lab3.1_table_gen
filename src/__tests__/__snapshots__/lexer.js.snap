// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lexer работаетб 1`] = `
[
  {
    "attr": "abba",
    "tag": "ident",
  },
  {
    "attr": null,
    "tag": "::=",
  },
  {
    "attr": null,
    "tag": ";",
  },
  {
    "attr": null,
    "tag": ",",
  },
  {
    "attr": null,
    "tag": "|",
  },
  {
    "attr": null,
    "tag": "|",
  },
  {
    "attr": "asdasd",
    "tag": "string",
  },
  {
    "attr": null,
    "tag": "#axiom",
  },
  {
    "attr": null,
    "tag": "#terminal",
  },
  {
    "attr": null,
    "tag": "#epsilon",
  },
  {
    "attr": null,
    "tag": "#non_terminal",
  },
  {
    "attr": null,
    "tag": "comment",
  },
]
`;
